Omnibar position option

0nko opened the pull request:
Task/Issue URL: https://app.asana.com/0/1207418217763355/1208356719296150/f

### Description

This is a feature branch for the omnibar position project.

### Testing instructions

#### Dummy toolbar

- [x] Start the omnibar at the top
- [x] Use the Fire button and notice the dummy toolbar is briefly shown at the top
- [x] Change the omnibar to the bottom
- [x] Use the Fire button and notice the dummy toolbar is briefly shown at the bottom

#### Autocomplete

- [x] Change the omnibar position to the bottom
- [x] Start typing in the address bar
- [x] Notice the suggestions are correctly displayed above the toolbar

#### UI changes
| Before  | After |
| ------ | ----- |
![image](https://github.com/user-attachments/assets/dfae5e47-a6e1-4d4d-bc25-4847ef4a0a32)|![image](https://github.com/user-attachments/assets/7ef7a0f9-cc74-41f5-8f28-d1d3462b740f)|

#### Snackbar

- [x] Change the omnibar position to the bottom
- [x] Trigger a condition to show a snackbar (see the diff below)
- [x] Notice the snackbar is displayed above the omnibar

This patch will triggers a snackbar after tapping the fire button:
```kotlin
Index: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt b/app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
--- a/app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt	(revision Staged)
+++ b/app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt	(date 1725012737445)
@@ -3299,6 +3299,7 @@
     }
 
     fun onFireMenuSelected() {
+        showErrorWithAction()
         val cta = currentCtaViewState().cta
         if (cta is OnboardingDaxDialogCta.DaxFireButtonCta) {
             onUserDismissedCta(cta)
``` 

#### UI changes
| Before  | After |
| ------ | ----- |
![image](https://github.com/user-attachments/assets/2f7c3158-48b9-4ffa-bb73-bd86fcba47f7)|![image](https://github.com/user-attachments/assets/76400958-f7d6-4ca7-b91b-9a9308807e74)|

#### Privacy pop-up dialog

- [x] Go to settings and change the omnibar position to BOTTOM
- [x] Go back to the tab
- [x] Trigger the privacy popup condition by refreshing the page
- [x] Notice the dialog correctly appears at the bottom

#### UI changes
| Before  | After |
| ------ | ----- |
![image](https://github.com/user-attachments/assets/b67b92ab-2b8e-428c-8f10-444c23f613c2)|![image](https://github.com/user-attachments/assets/c409136e-1986-4844-bc68-3ccfe7c4c593)|

#### Onboarding messages

- [x] Clear the app storage and trigger onboarding
- [x] Change the omnibar position to the bottom
- [x] Go to “google.com"
- [x] Notice that during onboarding steps, whenever there is the “hand” icon, it points to the bottom

#### UI changes
| Before  | After |
| ------ | ----- |
![image](https://github.com/user-attachments/assets/f4bad558-02d6-48cb-bea9-ab028d88b8b5)|![image](https://github.com/user-attachments/assets/49c4e50d-037a-45cb-922d-ac36b655edb2)|

#### Feature flag tests

_Bottom bar enabled_

Use `https://www.jsonblob.com/api/1287776240176848896` privacy config URL and the `playRelease` build, so that the feature flag is enabled

- [x] Go to Settings -> Appearance
- [x] Verify that there is the `Address bar` option to change the omnibar position

_Bottom bar disabled_

Use `http://www.jsonblob.com/api/1287777925473361920` privacy config URL, so that the feature flag is enabled

- [x] Go to Settings -> Appearance
- [x] Verify that the `Address bar` option is not visible

0nko commented:
* **#4885** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4885?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @0nko and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

malmstein reviewed (changes requested):
There are a couple of outstanding comments from previous PR. 
The big issue is that when the omnibar is at the bottom we are over the web content, and not properly pushing it to the top. That makes it so any content at the bottom of the viewport is behind the omnibar.

0nko commented on browser-api/src/main/java/com/duckduckgo/app/browser/omnibar/model/OmnibarPosition.kt:
```diff
@@ -0,0 +1,21 @@
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.omnibar.model
```
Based on what I observed, it seemed like the right place to put a common model that might be accessed by different modules in the future. If you feel it’s not the right place I’m open to suggestions.

malmstein replied:
If another module would need to access the omnibar position, they will do it via the `SettingsDataStore`. You have placed the model in the `browser-api` module, but the public accessor to that value is in the app module.

This module is specifically for things related to the browser. The omnibar and it’s position are separate from that. In other words, omnibar doesn’t need to know anything about the browser. I’d move it to the same app/omnibar package you have the rest of the logic.

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -3590,11 +3594,14 @@ class BrowserTabFragment :
         }
 
         private fun launchTopAnchoredPopupMenu() {
```
Fixed [here](https://github.com/duckduckgo/Android/pull/4885/commits/6b47869555c9fcbae445f0c5e5c3dd454634dea6).

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -3590,11 +3594,14 @@ class BrowserTabFragment :
-            } else {
-                pixel.fire(AppPixelName.MENU_ACTION_POPUP_OPENED.pixelName)
+            // small delay added to let keyboard disappear and avoid jarring transition
+            binding.rootView.postDelayed(100) {
```
Fixed [here](https://github.com/duckduckgo/Android/pull/4885/commits/a877ebfc3117e69a5a02946e38938aa3fcaf1ef9). Also fixed replaced the other literals [here](https://github.com/duckduckgo/Android/pull/4885/commits/db4c11a549658eb2ee4b5e741418488e8a5b25e4).

0nko commented:
> The big issue is that when the omnibar is at the bottom we are over the web content, and not properly pushing it to the top. That makes it so any content at the bottom of the viewport is behind the omnibar.

@malmstein Thanks for the review. I will fix this in a separate PR. The rest of the comments should be addressed now.

malmstein commented:
* **#4885** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4885?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 4  dependent PRs ([#4886](https://github.com/duckduckgo/Android/pull/4886) <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4886?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#4940](https://github.com/duckduckgo/Android/pull/4940) <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4940?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#4941](https://github.com/duckduckgo/Android/pull/4941) <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4941?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> and 1 other) 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @0nko and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -1165,6 +1168,8 @@ class BrowserTabViewModel @Inject constructor(
 
         canAutofillSelectCredentialsDialogCanAutomaticallyShow = true
 
+        onLoadProgressChanged(progress = newWebNavigationState.progress ?: 0)
```
this should be in `fun progressChanged()` and checking that progress is 100.

0nko replied:
Didn’t know about this method. I’ll move it there.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -1186,6 +1191,9 @@ class BrowserTabViewModel @Inject constructor(
                 }
+
+                // new page should always show an omnibar
+                command.value = ShowOmnibar
```
why is this needed?

0nko replied:
It fixes https://app.asana.com/0/1207418217763355/1208316032398497/f.

malmstein replied:
I asked @joshliebe to look at this too, it might affect the loading bar experiment he’s running.

joshliebe replied:
@0nko I tested this and the experiment works fine if `ChangeOmnibarPositionFeature` isn’t enabled. But when the loading bar is at the bottom there is always a delay, regardless of the whether the experiment is enabled or not.

You can test this by:

- Removing this from `LoadingBarExperimentVariantInitializer`:
```
loadingBarExperimentFeature.self().isEnabled() &&
loadingBarExperimentFeature.allocateVariants().isEnabled()
```
- And changing this line `val probabilities = doubleArrayOf(0.0, 1.0)` in `generateRandomBoolean()` so it becomes the test variant
- And finally, changing “Network download rate limit” in Android developer settings to the slowest and clicking a link.
- When the loading bar is at the top there is never a delay when a link is clicked, when it’s at the bottom there is a delay.

0nko replied:
@joshliebe Should be fixed now [here](https://github.com/duckduckgo/Android/pull/4885/commits/2714c263b1129262f72b4dbd9f78e2df9fad1159).

joshliebe replied:
@0nko I tested this and it’s working fine for the test variant but it’s also showing immediately for the control variant now (Which still needs the delay like before). This is happening if it’s at the bottom or top.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -1218,6 +1226,12 @@ class BrowserTabViewModel @Inject constructor(
     }
 
+    private fun onLoadProgressChanged(progress: Int) {
+        if (progress == 100 && settingsDataStore.omnibarPosition == BOTTOM) {
```
ditto, this belongs to `progressChanged()`

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/omnibar/Omnibar.kt:
```diff
@@ -0,0 +1,95 @@
+
+    val appBarLayout: LegacyOmnibarView by lazy {
+        if (omnibarPosition == OmnibarPosition.TOP || !omnibarPositionFeature.self().isEnabled()) {
+            binding.rootView.removeView(binding.legacyOmnibarBottom)
```
why remove the view instead of making it gone?

0nko replied:
This is a remnant of the previous implementation when we used included layouts. Both top and bottom would use the same view IDs so one had to be removed. I’ll make the change to hide it.

0nko replied:
Actually, after some testing it turns out this messes up the top omnibar. I think there are still ID conflicts, because it works fine when the views are removed. Reverting.

https://github.com/user-attachments/assets/06f91d82-fae4-4fc0-a67e-b86423cab981

malmstein commented on app/src/main/res/drawable/ic_autocomplete_down_24dp.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="20">
+  <path
+      android:pathData="M6.06,15h9.19a0.75,0.75 0,0 1,0 1.5h-11a0.75,0.75 0,0 1,-0.75 -0.75v-11a0.75,0.75 0,0 1,1.5 0v9.19L15.22,3.72a0.75,0.75 0,1 1,1.06 1.06L6.06,15"
+      android:fillColor="#000"/>
```
```suggestion
            android:fillColor="?attr/daxColorPrimaryIcon"/>
```

this needs to be changed, otherwise it will show black in dark mode.

malmstein commented on app/src/main/res/layout/fragment_browser_tab.xml:
```diff
@@ -124,4 +120,13 @@
 
     </RelativeLayout>
+
+    <com.duckduckgo.app.browser.omnibar.LegacyOmnibarView
```
this could be gone by default

malmstein commented on app/src/main/res/layout/include_omnibar_toolbar_mockup.xml:
```diff
@@ -16,111 +16,101 @@
+        android:id="@+id/tabsMenuMockup"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginHorizontal="@dimen/keyline_2"
```
are these changes intentional? I was under the assumption that there wouldn’t be padding changes in this PR

0nko replied:
This just fixes the alignment issue with the mock toolbar. The padding changes will come in a different PR.

malmstein commented on app/src/main/res/layout/view_legacy_omnibar.xml:
```diff
@@ -15,303 +15,300 @@
-                android:layout_marginStart="@dimen/keyline_2"
-                android:layout_marginEnd="@dimen/keyline_2"
-                android:background="?selectableItemBackgroundBorderless"
+                android:layout_gravity="center"
```
more padding changes. is this supposed to be here?

0nko replied:
See my previous comment.

malmstein reviewed (commented):
- There are padding changes to the Omnibar that I was under the impression we weren’t going to add.
- The autocomplete icon can’t be seen in dark mode. It’s black, when it should be using the daxColor for icons so dark theme is respected.
- You shouldn’t check the site progress in a navigation change, but in `progressChanged()`.
- There’s a new Command that forces the Omnibar to be shown on navigation change. This wasn’t here before, why is this needed?

0nko commented:
@malmstein I’ve addressed your comments but I also had to make some additional changes:

- I changed the way we check if the feature flag is enabled. There are many places that check the position value. To avoid adding a feature flag check everywhere, I added a [plugin](https://github.com/duckduckgo/Android/pull/4885/commits/5001ec5b2410cc952aea5072d0848c86eb973039#diff-62cb0adbb8cc92c060ba94ecab5e7f220ae563d2367c5210e85fa846182e6b10) that checks on process start and modifies the data store.
- I updated the JS that checks the scrollability of a website and it’s now able to correctly determine that `theguardian.com` and similar sites have a non-scrollable elements at the front
- I found that the omnibar is scrollable when showing autocomplete (a behavior that is present in production) and I made it fixed

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -943,38 +950,93 @@ class BrowserTabFragment :
+        }
+    }
+
+    private suspend fun checkIfSiteScrollable(duckDuckGoWebView: DuckDuckGoWebView) {
```
nit: this method not only checks for scrollable, it also changes paddings and layout params. The method name should reflect that too.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/omnibar/OmnibarFeatureFlagObserver.kt:
```diff
@@ -0,0 +1,54 @@
+    boundType = MainProcessLifecycleObserver::class,
+)
+@SingleInstanceIn(AppScope::class)
+class OmnibarFeatureFlagObserver @Inject constructor(
```
This is not the right place to do this. On process start will get called before a privacy config has been downloaded. If the feature flag value in remote has changed, you won’t get the `settingsDataStore` updated until the next time the app cold starts.

What you want to use is `PrivacyConfigCallbackPlugin`. This plugin will notify you when the privacy config has been downloaded, and will let you update the position if necessary.

Check https://github.com/duckduckgo/Android/blob/develop/experiments/experiments-impl/src/main/java/com/duckduckgo/experiments/impl/loadingbarexperiment/LoadingBarExperimentVariantInitializer.kt as an example.

0nko replied:
Ah, good to know, thanks. I’ve fixed it [here](https://github.com/duckduckgo/Android/pull/4885/commits/a9fa84366fa19b58df6f8029c98e084d29f3941f).

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/omnibar/TopAppBarBehavior.kt:
```diff
@@ -0,0 +1,44 @@
+ * This custom behavior for the bottom omnibar is necessary because the default `HideBottomViewOnScrollBehavior` does not work.
+ * The reason is that the `DuckDuckGoWebView` is passing only unconsumed movement, which `HideBottomViewOnScrollBehavior` ignores.
+ */
+class TopAppBarBehavior(context: Context, attrs: AttributeSet?) : AppBarLayout.Behavior(context, attrs) {
```
Why is this needed? The description mentions bottom omnibar but this is applied when the bar is at the top.

0nko replied:
I forgot to update the comment. This is needed to fix the scrollable top bar during autocomplete:

> I found that the omnibar is scrollable when showing autocomplete (a behavior that is present in production) and I made it fixed

0nko replied:
Fixed.

malmstein reviewed (changes requested):
The feature flag won’t be updated properly with this code. Please take a look at the comment and we can go over it if needed.

aitorvs commented on app/src/main/java/com/duckduckgo/app/browser/omnibar/ChangeOmnibarPositionFeature.kt:
```diff
@@ -0,0 +1,33 @@
+    featureName = "changeOmnibarPosition",
+)
+interface ChangeOmnibarPositionFeature {
+    @Toggle.DefaultValue(false)
+    fun self(): Toggle
```
I saw you created the config set to `internal`. You don't really need the config just yet, you could just annotate this method with `@InternalAlwaysEnabled`

0nko commented:
@malmstein I’ve updated the test instructions to include tests of the feature flag, as well.

malmstein reviewed (approved):
Looking good and works as expected. Thanks @0nko !
